src/datadog/telemetry/telemetry_impl.cpp

```diff
@@ -244,27 +354,339 @@ void Telemetry::send_telemetry(StringView request_type, std::string payload) {
+            {"interval", 10},
+            {"points", points},
+            {"common", metric.common()},
+        }));
+      }
+    }
+    points.clear();
+  }
+
+  if (!metrics.empty()) {
+    auto generate_metrics = nlohmann::json::object({
+        {"request_type", "generate-metrics"},
+        {"payload", nlohmann::json::object({
+                        {"series", metrics},
+                    })},
+    });
+    batch_payloads.emplace_back(std::move(generate_metrics));
+  }
+
+  if (!logs_.empty()) {
+    auto encoded_logs = nlohmann::json::array();
+    for (const auto& log : logs_) {
+      auto encoded = encode_log(log);
+      encoded_logs.emplace_back(std::move(encoded));
+    }
+
+    assert(!encoded_logs.empty());
+
+    auto logs_payload = nlohmann::json::object({
+        {"request_type", "logs"},
+        {"payload",
+         nlohmann::json{
+             {"logs", encoded_logs},
+         }},
+    });
+
+    batch_payloads.emplace_back(std::move(logs_payload));
+  }
+
+  auto telemetry_body = generate_telemetry_body("message-batch");
+  telemetry_body["payload"] = batch_payloads;
+  auto message_batch_payload = telemetry_body.dump();
+
+  return message_batch_payload;
+}
+
+std::string Telemetry::app_closing() {
+  auto batch_payloads = nlohmann::json::array();
+
+  auto app_closing = nlohmann::json::object({
+      {"request_type", "app-closing"},
+  });
+  batch_payloads.emplace_back(std::move(app_closing));
+
+  auto metrics = nlohmann::json::array();
+  for (auto& m : metrics_snapshots_) {
+    auto& metric = m.first.get();
+    auto& points = m.second;
+    if (!points.empty()) {
+      auto type = metric.type();
+      if (type == "count") {
+        metrics.emplace_back(nlohmann::json::object({
+            {"metric", metric.name()},
+            {"tags", metric.tags()},
+            {"type", metric.type()},
+            {"points", points},
+            {"common", metric.common()},
+            {"namespace", metric.scope()},
+        }));
+      } else if (type == "gauge") {
+        // gauge metrics have a interval
+        metrics.emplace_back(nlohmann::json::object({
+            {"metric", metric.name()},
+            {"tags", metric.tags()},
+            {"type", metric.type()},
+            {"interval", 10},
+            {"points", points},
+            {"common", metric.common()},
+            {"namespace", metric.scope()},
+        }));
+      }
+    }
+    points.clear();
+  }
+
+  if (!metrics.empty()) {
+    auto generate_metrics = nlohmann::json::object({
+        {"request_type", "generate-metrics"},
+        {"payload", nlohmann::json::object({
+                        {"series", metrics},
+                    })},
+    });
+    batch_payloads.emplace_back(std::move(generate_metrics));
+  }
+
+  if (!logs_.empty()) {
+    auto encoded_logs = nlohmann::json::array();
+    for (const auto& log : logs_) {
+      auto encoded = encode_log(log);
+      encoded_logs.emplace_back(std::move(encoded));
+    }
```

Comment:
This function body looks very similar to the above function body. Would it be worth to refactor that ?

Reply:
That's planned but not now.